### PR TITLE
Mobt 496 enforce forecast between references

### PR DESCRIPTION
**Pull request by brhooper**
Addresses #1900 
Acceptance test data: https://github.com/metoppv/improver_test_data/pull/12

Adds functionality to enforce forecast to be within lower and upper bounds derived from linear functions of a reference forecast. This extends functionality implemented in #1900 .

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Review by Katie-Howard — CHANGES_REQUESTED** [review]
I've had a look through these changes, and  they generally look fine. I just have one question to be addressed. I've also had a little look at the acceptance test data and left a comment on that too. I couldn't see any data in the new file, so if you could check that please?

**Review by Katie-Howard — APPROVED** [review]
These changes all look fine. I reran all the tests and they passed fine 👍

**Review by mspelman07 — CHANGES_REQUESTED** [review]
Thanks Ben. Generally looks good, just a few small queries.

File: improver/cli/enforce_consistent_forecasts.py [hunk 1]
```diff
@@ -39,14 +39,14 @@
 def process(
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/cli/enforce_consistent_forecasts.py:39]: Removed now, thanks for pointing this out

File: improver/utilities/enforce_consistency.py [hunk 2]
```diff
@@ -81,69 +86,129 @@ def __init__(
...
-            enforced to be not less than or not greater than a linear function of
-            reference_forecast.
+            enforced to be within the calculated bounds.
         """
+
         # check forecast and reference units match
         if forecast.units != reference_forecast.units:
             msg = (
-                f"The units in the forecast and reference cubes do not match. The"
+                "The units in the forecast and reference cubes do not match. The"
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/utilities/enforce_consistency.py]: Added a `try` ... `except` block to attempt to convert units, raising the existing error if an exception is raised.

File: improver/utilities/enforce_consistency.py [hunk 3]
```diff
@@ -81,69 +86,129 @@ def __init__(
...
                 f"units of forecast were {forecast.units}, the units of "
                 f"reference_forecast were {reference_forecast.units}."
             )
             raise ValueError(msg)
 
         # linear transformation cannot be applied to probability forecasts
-        if self.additive_amount is not None or self.multiplicative_amount is not None:
+        if self.additive_amount != 0.0 or self.multiplicative_amount != 1.0:
             if forecast.name().startswith("probability_of"):
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/utilities/enforce_consistency.py]: Thanks for highlighting this, I was looking for just such a function, without success, when initially making the changes, so this is very helpful

File: improver/utilities/enforce_consistency.py [hunk 4]
```diff
@@ -81,69 +86,129 @@ def __init__(
...
+
         # check forecast and reference units match
         if forecast.units != reference_forecast.units:
             msg = (
-                f"The units in the forecast and reference cubes do not match. The"
+                "The units in the forecast and reference cubes do not match. The"
                 f"units of forecast were {forecast.units}, the units of "
                 f"reference_forecast were {reference_forecast.units}."
             )
             raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/utilities/enforce_consistency.py]: Added to the docstring

**Review by mspelman07 — CHANGES_REQUESTED** [review]
All the changes look good, just one tiny thing and I'm happy to approve.

File: improver/utilities/enforce_consistency.py [hunk 5]
```diff
@@ -81,69 +87,147 @@ def __init__(
...
+
+        Raises:
+            ValueError: If units of forecast and reference cubes are different and
+                cannot be converted to match.
+            ValueError: If additive_amount and multiplicative_amount are not 0.0 and 1.0,
+                respectively, when a probability forecast is input.
+            ValueError: If incorrect comparison_operator is input.
+            ValueError: If contradictory bounds are generated.
+            ValueError: If any of additive_amount, multiplicative_amount, or
+                comparison_operator are lists when ther are not all lists.
```
[the remark this answers is not in the document]
Comment (brhooper) [reply improver/utilities/enforce_consistency.py]: Corrected

**Review by mspelman07 — APPROVED** [review]

**Pull request merged**